### PR TITLE
HDDS-8897. Avoid synchronization on WritableECContainerProvider

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

Keep track of pending EC pipeline allocations in `WritableECContainerProvider` to avoid the need for synchronization.  Only force wait for over-the-limit requests.

https://issues.apache.org/jira/browse/HDDS-8897

## How was this patch tested?

https://github.com/adoroszlai/hadoop-ozone/actions/runs/5342367241

### CODE REVIEW

**Comment by adoroszlai** [comment]
@guohao-rosicky please check if this helps improve latency

**Comment by sodonnel** [comment]
We need to see a profile of the original code and the changed code to see exactly where the problem lies. There is synchronization inside pipeline and container manager too, so the changes here might not help much, as we are also not certain about where the problem is, as there are several synchronized blocks in the original code.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableECContainerProvider.java [hunk 1]
```diff
@@ -150,21 +153,70 @@ public ContainerInfo getContainer(final long size,
...
+
+    final int pending = pendingAllocations.getAndIncrement();
     try {
-      synchronized (this) {
-        if (openPipelineCount < maximumPipelines) {
-          return allocateContainer(repConfig, size, owner, excludeList);
+      if (current + pending < max) {
+        ContainerInfo containerInfo =
+            allocateContainer(repConfig, size, owner, excludeList);
+        allocation.signal();
```
Comment (sodonnel) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/WritableECContainerProvider.java:191]: Is there a potential race condition here?

Say we are 1 off the max, and 1 pipline is pending create.

Another thread falls into the "else if". Just before it calls `allocation.await()` the previous allocation completes and calls allocation.signal().

Then the "other thread" will get blocked on allocation.await() with nothing to wake it up?

**Comment by guohao-rosicky** [comment]
hi @adoroszlai 
I am also following this issue, this lock contention will degrade the performance of allocate block, detailed report I have submitted to jira. 
please see: https://issues.apache.org/jira/browse/HDDS-8897

**Comment by guohao-rosicky** [comment]
@adoroszlai
I have an idea to asynchronously process the close pipeline and pre-creating the container, so that the lock time will be much shorter.

I now have a version of the initial implementation of the code, whether to submit a pr, according to the code we will discuss

**Comment by sodonnel** [comment]
@guohao-rosicky We really need to profile the code to see the hot spots before jumping into solutions. As it stands we don't know:

1. If the problem is too few pipelines - see my comment on the Jira about this.
2. Around pipelines creation time (which could be made worse by too few pipelines)
3. Selecting pipelines from the open set 

If you could reproduce your test with the async profiler attached to SCM, it would be a great help. If it difficult to profile this in a simulation, as the need to open pipelines is driven by pipelines filling and closing, so it really needs a workload on a real cluster.

**Pull request closed without merging**